numpyro/util.py

```diff
@@ -164,6 +165,70 @@ def _wrapped(fn):
     return _wrapped
 
 
+def progress_bar_factory(num_samples):
+    """Factory that builds a progress bar decorator along
+    with the `set_tqdm_description` and `close_tqdm` functions
+    """
+
+    tqdm_bars = {}
+
+    def _define_tqdm(arg, transform, device):
+        chain = int(str(device)[4:])
+        tqdm_bars[chain] = tqdm.tqdm(range(num_samples))
```

Comment:
By the way, I guess we need to use `from tqdm.auto import tqdm` here. Using `tqdm.tqdm` does not display the bars well in my notebook. But probably it is just my issue.

Reply:
I tried adding `host_callback.barrier_wait()` just after the `fori_loop()` in `fori_collect` but the sampler simply doesn't run anymore.. :(. 

The issue seems fixed for notebook (at least when I tried it) though which is good. When running it in the terminal it can still look a bit messy though

Reply:
How about adding `host_callback.barrier_wait()` at the end of mcmc.run (when num_chains > 1, chain_method="parallel", and progress_bar=True)? I guess it might do the trick.

Reply:
Sorry for takes ages to get to this!

I don't quite get this bit: `progress_bar=True` is for the progress bar for the un-compiled chain, no?

Reply:
Yes, that position would work IMO. I think as long as it is outside of `jax.pmap`, it will be fine. Probably, after [this line](https://github.com/pyro-ppl/numpyro/blob/ce6cac87ec715ea467a353160dcb56a1827b28a9/numpyro/infer/mcmc.py#L504) is better...

Reply:
I tried added it where you suggested and the extra progress bars still appear (when running a script in the terminal). Do you get the same behaviour?

Reply:
It seems that the following code also works in my system
```python
def progress_bar_factory(num_samples, num_chains):
    ...
    finished_chains = []
    for chain in range(num_chains):
        tqdm_bars[chain] = tqdm_auto(range(num_samples), position=chain)
        message = f"Running chain {chain}"
        tqdm_bars[chain].set_description(message, refresh=False,)
        tqdm_bars[chain].update(0)

    def _close_tqdm(arg, transform, device):
        chain = int(str(device)[4:])
        tqdm_bars[chain].update(arg)
        finished_chains.append(chain)
        if len(finished_chains) == num_chains:
            for chain in range(num_chains):
                tqdm_bars[chain].close()

    def _update_progress_bar(iter_num):
        _ = lax.cond(
            iter_num % print_rate == 0,
            lambda _: host_callback.id_tap(_update_tqdm, print_rate, result=iter_num, tap_with_device=True),
            lambda _: iter_num,
            operand=None,
        )
        _ = lax.cond(
            iter_num == num_samples,
            lambda _: host_callback.id_tap(_close_tqdm, remainder, result=iter_num, tap_with_device=True),
            lambda _: iter_num,
            operand=None,
        )

    def progress_bar_fori_loop(func):
        def wrapper_progress_bar(i, vals):
            result = func(i, vals)
            _update_progress_bar(i + 1)
            return result
```
The changes that I made are:
+ start the chains in order: this is achieved by moving `_define_tqdm` out of the fori loop. Without this, the progress bars do not appear in order.
+ close the chains in order: this is achieved by adding a variable `finished_chains` to record which chains have been finished. After all chains are finished, we close all progress bars. Without this, in the terminal, the progress bars swap together in an arbitrary order (probably the chain that finished first will be on the top - I don't know). 
+ report the progress only after the current iteration finished its job. This is a bit different from the current behavior in your code, where progress bars are updated at positions 0, print_rate, 2 * print_rate (while I think it should be print_rate, 2*print_rate,...). Can we discuss a bit more on what is a better behavior here?

Reply:
I added the code you suggested:
- Updating the progress bar after running the iterations makes sense!
- On my system: your fix means that the chains now appear in order. I also now don’t have an extra progress bar appearing in the middle of a run and then disappearing.
- I tested different number of samples/chains/warmup: this seems to work fine
- The `barrier_wait()` is not needed on my system when running your MH example. Which bits of texts were badly positioned for you?
- For me the only small point left is: The chains now start at 0% and wait for the sampler to compile before running. Perhaps a message `print(“Compiling..”)` at the beginning of the progress bar would be useful to make it clear that everything is working fine. However this is not really a problem to be honest, and perhaps printing too much stuff isn't great either.

> Could you provide the script so I can test it in my system?

I've been using the 8 schools example to test my code throughout this PR:

```
import numpy as np
import time

J = 8
y = np.array([28.0, 8.0, -3.0, 7.0, -1.0, 1.0, 18.0, 12.0])
sigma = np.array([15.0, 10.0, 16.0, 11.0, 9.0, 11.0, 10.0, 18.0])

import numpyro
import numpyro.distributions as dist
numpyro.set_host_device_count(4)

def eight_schools(J, sigma, y=None):
    mu = numpyro.sample('mu', dist.Normal(0, 5))
    tau = numpyro.sample('tau', dist.HalfCauchy(5))
    with numpyro.plate('J', J):
        theta = numpyro.sample('theta', dist.Normal(mu, tau))
        numpyro.sample('obs', dist.Normal(theta, sigma), obs=y)


from jax import random
from numpyro.infer import MCMC, NUTS

print("Define sampler")
nuts_kernel = NUTS(eight_schools)

mcmc = MCMC(nuts_kernel, num_warmup=502, num_samples=100000, progress_bar=True, num_chains=4)
rng_key = random.PRNGKey(0)

starttime = time.time()
mcmc.run(rng_key, J, sigma, y=y, extra_fields=('potential_energy',))
endtime = time.time()
print(f"Running time: {endtime-starttime:.1f}sec")


mcmc.print_summary()
```

Reply:
> Which bits of texts were badly positioned for you?

Nevermind, in my system, sometimes the line `abccc` appears in the middle of the bars (in terminal). It is pretty rare and only happens for very fast `mcmc.run`. I would prefer to skip adding `barrier_wait`, which I don't fully understand its effect.

> The chains now start at 0% and wait for the sampler to compile before running

Understood. We can start the bar with "Compiling..." (also sample the first step) and switch to the current sampling text when `i == 1`. Please feel free to adjust this if you want.

Reply:
I added the "Compiling.." messages in the tqdm progress bars which then get overridden when sampling starts. Is this what you meant?

Reply:
Sorry, I meant to add
```python
    def _update_tqdm(arg, transform, device):
        chain = int(str(device)[4:])
        tqdm_bars[chain].set_description(f"Running chain {chain}", refresh=False,)
        tqdm_bars[chain].update(arg)

    def _update_progress_bar(iter_num):
        _ = lax.cond(
            iter_num == 1,
            lambda _: host_callback.id_tap(_update_tqdm, 0, result=iter_num, tap_with_device=True),
            lambda _: iter_num,
            operand=None,
        )
```
and starting with `Compiling..` as you did. It seems that `python examples/hmm.py --num-chains 4` gives expected behavior.

Reply:
I tried this solution and it seems that both methods give the same behaviour (for both the hmm.py example and the 8 schools example). 

Is this the same for you? And do you have a preference for which one we use?

Reply:
IIUC this will keep Compiling.. until 5% samples are collected. I think it is better to keep Compiling.. until the first sample is collected. WDYT?

Reply:
Ah yeah you're right!
I just fixed it; the message now appears at the start of sampling.